File: leetcode/1186. Maximum Subarray Sum with One Deletion.py
class Solution:
    def maximumSum(self, arr: List[int]) -> int:
        dp = [[0, 0] for _ in range(len(arr))]
        res = -sys.maxsize
        
        for i, num in enumerate(arr):
            if i != 0:
                last = dp[i-1][0]
                last_d = dp[i-1][1]
            else:
                last = -sys.maxsize
                last_d = -sys.maxsize

            dp[i][0] = max(arr[i], last+arr[i])
            dp[i][1] = max(arr[i], last, last_d+arr[i])
            res = max(res, dp[i][0], dp[i][1])
        
        return res

File: leetcode/1186. Maximum Subarray Sum with One Deletion.cpp
class Solution {
public:
    int maximumSum(vector<int>& arr) {
        vector<vector<int>> dp(arr.size(), vector<int>(2, 0));

        dp[0][0] = arr[0]; // not delete
        dp[0][1] = 0; // deleted
    
        int res = dp[0][0];
        int max_num = arr[0];
        
        for (size_t i = 1; i < arr.size(); ++i) {
            dp[i][0] = max(dp[i-1][0], 0) + arr[i];
            dp[i][1] = max({dp[i-1][1] + arr[i], dp[i-1][0], 0});
            res = max({res, dp[i][0], dp[i][1]});
            max_num = max(max_num, arr[i]);
        }
        
        if (max_num < 0) {
            return max_num;
        }

        return res;
    }
};
